charts/landscaper/gardenlet/vpa-crd/Chart.yaml

```diff
@@ -0,0 +1,4 @@
+apiVersion: v1
```

Comment:
Why do we have this in our respository? VPA is a prerequisite in the runtime cluster. Let's not maintain this CRD here.

Reply:
> VPA is a prerequisite in the runtime cluster.

where is this from and why should that be the case?


Regardless this component should be used to register a  Kubernetes cluster as a Seed resource. This can be any cluster. 
Why would we define something as a prerequisite for an automated installer? In my point of view the point of the installer is  to everything thats needed (^not have manual steps)

Reply:
From your point of view, why do you then only deploy the CRD (actually, you only deploy one of the two VPA CRDs) and not the whole VPA components, RBAC roles, and everything that comes with it then, including the exposure of configuration options?
With the same line of thought, why don't you deploy the Kubernetes cluster yourself into which you later deploy the gardenlet? Instead, you just take a kubeconfig and let some(body/ other component) manage the runtime cluster. With the VPA and the RBAC roles it's similarly. 
https://github.com/gardener/gardener/pull/3130#discussion_r523926747 clearly indicates that it's wrong to put those things here because you cannot fully control the lifecycle of those dependencies - it's out of scope of this component.
Let's please not make the same mistakes than with the old landscape-setup and rather design the components with a clear scope and API.

Reply:
> From your point of view, why do you then only deploy the CRD (actually, you only deploy one of the two VPA CRDs) and not the whole VPA components, RBAC roles, and everything that comes with it then, including the exposure of configuration options?

I though that the VPA CRD is the only resource missing in the Seed cluster. If there is more, sure, I think we should add it.

> With the same line of thought, why don't you deploy the Kubernetes cluster yourself into which you later deploy the gardenlet? 

Because the goal is to register a Kubernetes cluster. Not to create one. 
Deploying a Kubernetes resource to prepare the Seed cluster is in my opinion not comparable to creating the cluster itself?.

> #3130 (comment) clearly indicates that it's wrong to put those things here because you cannot fully control the lifecycle of those dependencies - it's out of scope of this component.

I agree that when removing the landscaper gardenlet component again, we cannot safely remove the VPA CRD (for the Gardenlet RBAC we actually can,  by checking if there are still Seeds in the system).
Is that a problem?

As an example: when the Gardenlet debootstraps the Seed cluster it does NOT delete the garden namespace because it cannot be safely deleted.
Based on your thoughts above  we should make the Garden namespace a prerequisite before bootstrapping a Seed cluster. 
However this is neither practical nor contributes to a good user experience.

Providing a streamlined Gardener installation is one of the major Goals of the landscaper.
If we continue on this path, I fear we ll end up with an installer that has many dependencies and manual  steps involved making it uncomfortable to use.

> Let's please not make the same mistakes than with the old landscape-setup and rather design the components with a clear scope and API.

How does deploying the VPA CRD  introduce additional complexity or increases the API surface of the landscaper component?. 

Coming back to the RBAC roles: I think having the RBAC roles required for the Gardenlet to run within the Gardenlet component contributes to that Goal. Not have it spread among other landscaper components.

Reply:
lets involve some more colleagues @timebertt @schrodit @BeckerMax @timuthy .
In the end I am fine with either ways. Just want to make sure that we involved the right people and make an informed decision.
Thanks @rfranzke

Reply:
> > Let's please not make the same mistakes than with the old landscape-setup and rather design the components with a clear scope and API.
> 
> How does deploying the VPA CRD introduce additional complexity or increases the API surface of the landscaper component?.

VPA is not only composed of the one CRD (as @rfranzke already stated). In order to get it to work, you also have to deploy the controllers, rbac, webhooks and so on. So this indeed introduces a lot of additional complexity. 
I agree with @rfranzke, that the gardenlet component should not be responsible for this additional complexity.

So I guess, we either have to state that VPA is a requirement for the runtime cluster or add a separate VPA component, that will deploy and manage it (can also be reused for other components that will require a VPA installation).

Reply:
> I agree that when removing the landscaper gardenlet component again, we cannot safely remove the VPA CRD 
Is that a problem?

Doesn't this fact already indicate that it doesn't belong into this component? Leaving orphaned resources usually is undesired and should be avoided.

> (for the Gardenlet RBAC we actually can, by checking if there are still Seeds in the system).

Well, sure, but that is a workaround that - by chance - might work. Still, I consider the RBAC resources part of the Gardener control plane (see below).

> As an example: when the Gardenlet debootstraps the Seed cluster it does NOT delete the garden namespace because it cannot be safely deleted.
Based on your thoughts above we should make the Garden namespace a prerequisite before bootstrapping a Seed cluster.
However this is neither practical nor contributes to a good user experience.

I don't understand. The gardenlet doesn't delete the `garden` namespace because it doesn't create it in the first place, or? Somebody must have created the `garden` namespace before deploying the gardenlet into it. Hence, this "somebody" is responsible for the namespace, not the gardenlet itself.

> Providing a streamlined Gardener installation is one of the major Goals of the landscaper.

That's the goal of the whole landscape-setup that is to be introduced, but it's not the goal of this particular component. Again, the composition of components is what will be delivered, and gardenlet is only one of many. It has a clear scope and should be limited to it.

> If we continue on this path, I fear we ll end up with an installer that has many dependencies and manual steps involved making it uncomfortable to use.

Definitely, there will be some prerequisites that must be fulfilled, but that's always the case if you want to install software. For this particular discussion, the VPA and the RBAC roles won't be required to be setup/deployed manually, of course. It's just that other landscaper components will be responsible for them. Not this one.

> How does deploying the VPA CRD introduce additional complexity or increases the API surface of the landscaper component?.

It might be "small", but it still does not belong into this component. Separation of concerns.

> Coming back to the RBAC roles: I think having the RBAC roles required for the Gardenlet to run within the Gardenlet component contributes to that Goal. Not have it spread among other landscaper components.

Disagree. To emphasise my argument, let's (as usual) take a look at how Kubernetes is doing it: Each cluster has a `system:node-bootstrapper` cluster role and a cluster role binding. Those RBAC resources are not created by individual kubelets, but by the kube-apiserver in its bootstrapping phase. With Gardener, it's similar: They are created as part of the control plane deployment (which includes the gardener-apiserver). If we want, we can even remove them from the Helm chart and do it similar to what the kube-apiserver is doing (creating them in the bootstrapping phase). Anyways, I hope this makes it more clear.

> So I guess, we either have to state that VPA is a requirement for the runtime cluster or add a separate VPA component, that will deploy and manage it (can also be reused for other components that will require a VPA installation).

Exactly @timebertt, that's it.